Improvements to hooks

afbora opened the pull request:
## Describe the PR

**General note:** As this is a large PR, reviewing this from the "Commits" view instead of from the "Files changed" view is recommended to keep the overview. I was very careful to split the PR up into useful commits that make sense on their own, which should help with reviewing in this way. Splitting this PR up into separate PRs was unfortunately not possible as we would then have needed three PRs at minimum, which would be even harder to keep an overview on.

- Hook arguments are now named and magically provided as requested (like in controllers). So it's now possible to just request the arguments the hook needs and in any order. Arguments that are not available will be set to `null`.
- Support for wildcard hooks (e.g. `page.*:before` or `page.create:*` or `*.create:*` – all variations are supported!) that will be triggered for all event actions of the given type.  
  *Note:* The support for all wildcard variations is included in the fifth commit. If we decide that a simple `page.*:before` hook without the other variations is enough (which wouldn't be as flexible though), we can simply remove the fifth commit.
- New `$event` object provided as argument for all hooks that contains all information about the triggered event. The `$event` argument can be used instead of or in addition to the individual hook arguments.

```php
return [
    'hooks' => [
        'page.*:before' => function ($page, $event) {
            var_dump($page);            // $page
            var_dump($event->name());   // 'page.create:before'
            var_dump($event->type());   // 'page'
            var_dump($event->action()); // 'create'
            var_dump($event->state());  // 'before'
            var_dump($event->page());   // $page
            var_dump($event->input());  // $input
        }
    ]
];
```

**Breaking-changes:**

- The method signature of `$app->apply()` and `$app->trigger()` has changed. Instead of a list of arguments, they now receive an associative array with named arguments. `$app->apply()` now also requires a third argument that tells the method which argument should be modified by the hooks. As both methods are marked for internal use and also not very useful in sites, the only use-case for them so far was hooks defined in plugins. Plugin authors who depend on any of the two changed methods need to update their plugins to support the new signature.
- As the hook arguments are now detected by name and not by order, any hooks that don't use the variable naming as [documented in the reference](https://getkirby.com/docs/reference/plugins/extensions/hooks#available-events) will receive `null` for those arguments instead. I expect that most users already use the naming from the reference as the reference provides copy-paste code snippets. However all hooks should be checked for wrong variable naming. Sometimes the issue will be obvious as the hook will just not work, but it's still better to check beforehand. I have developed a [migration checker plugin](https://github.com/getkirby/kirby/pull/2541#issuecomment-613467839) for this.

## Related issues

<!-- PR relates to issues in the `kirby` or `idea` repo: -->

- Implements https://github.com/getkirby/ideas/issues/194

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Fixed code style issues with CS fixer and `composer fix`
- [x] Added in-code documentation (if needed)

lukasbestle reviewed (commented):
I agree with Neil that the current implementation is a bit too complex and does too much. I think we should keep this as simple as possible for now to make it easier to understand. Which means that the `Hook` should probably only have the methods it absolutely needs.

Regarding implementation:

- Now that I think about it, I'm not sure about the different arguments for "old-style" hooks and the new global hooks. What we could do is to use `ReflectionFunction` like in the `Toolkit\Controller` class to determine at run-time which arguments each individual hook would like to receive. We could even support arbitrary argument orders that way and it would also be possible to receive `$hook` *in addition* to some or all of the individual arguments.
- Triggering both the individual as well as the "global" hooks should be automated in the `$app->trigger()` method. So `$app->trigger()` would ideally only take a `Hook` object and then determine the hooks to call from that.
- I'm not sure how we can make this feature work with `$app->apply()`. To make it all consistent, everything we change about `$app->trigger()` should also be changed for `$app->apply()`. But that's going to be a bit difficult.

lukasbestle committed:
$app->trigger()/->apply(): Use Event class

Allows for arbitrary argument orders in hook functions as well as for a new optional `$event` argument for every hook.

NOTE: Breaking-change as the method signature of $app->trigger() and $app->apply() has changed.

Implements https://github.com/getkirby/ideas/issues/194.

lukasbestle committed:
Adapt core code to trigger()/apply() changes

https://github.com/getkirby/ideas/issues/194

lukasbestle committed:
New Kirby\Cms\Event class

https://github.com/getkirby/ideas/issues/194

lukasbestle committed:
Add support for wildcard hooks

https://github.com/getkirby/ideas/issues/194

lukasbestle committed:
Wildcard hooks: Full wildcard variation support

https://github.com/getkirby/ideas/issues/194

distantnative commented on src/Cms/App.php:
```diff
@@ -184,37 +184,42 @@ public function api()
      * @return mixed Resulting value as modified by the hooks
      */
-    public function apply(string $name, ...$args)
+    public function apply(string $name, array $args, string $modify, ?Event $originalEvent = null)
```
@lukasbestle Just a quick note: We wanted to stay away from type hinting objects

lukasbestle replied:
Wasn't that just about method return values though? We have quite a few methods that use class type hinting for arguments, for example in the `AppPlugins` trait. Also I think it's not a problem for `App` as that class is not going to be extended by plugins.

But if you think we should remove the type hinting, that's no problem either.

lukasbestle replied:
@distantnative Ping :)
As I said, either way is fine for me, but we need to decide.

afbora commented on src/Cms/App.php:
```diff
@@ -184,37 +184,42 @@ public function api()
+     * @param string $name Full event name
+     * @param array $args Associative array of named event arguments
+     * @param string $modify Key in $args that is modified by the hooks
+     * @param \Kirby\Cms\Event|null $originalName Event object (internal use)
```
```suggestion
     * @param \Kirby\Cms\Event|null $originalEvent Event object (internal use)
```

afbora commented on src/Cms/App.php:
```diff
@@ -1314,12 +1319,15 @@ public function thumb(string $src, string $dst, array $options = []): string
-     * @param mixed ...$arguments
+     * @param string $name Full event name
+     * @param array $args Associative array of named event arguments
+     * @param \Kirby\Cms\Event|null $originalName Event object (internal use)
```
```suggestion
     * @param \Kirby\Cms\Event|null $originalEvent Event object (internal use)
```

lukasbestle committed:
Fix docblocks

Co-Authored-By: Ahmet Bora <byybora@gmail.com>

bastianallgeier committed:
Merge branch 'develop' into feature/194-wildcard-hooks
